server/src/main/java/org/elasticsearch/search/internal/ContextIndexSearcher.java

```diff
@@ -226,9 +240,15 @@ public Weight createWeight(Query query, ScoreMode scoreMode, float boost) throws
         }
     }
 
+    /**
+     * Overwrite superclass to force one slice per segment for knn search.
+     * This is only needed temporarily by knn query rewrite, for the main
+     * search collection we forked the search method and inject our own slicing logic
+     * until this is available in Lucene itself
+     */
     @Override
     protected LeafSlice[] slices(List<LeafReaderContext> leaves) {
-        return computeSlices(leaves, queueSizeBasedExecutor.threadPoolExecutor.getMaximumPoolSize(), minimumDocsPerSlice);
+        return IndexSearcher.slices(leaves, Math.max(1, leaves.size()), 1);
```

Comment:
Could you add a unit test to ContextIndexSearcherTests that verifies the discrepancy between the two slices method, and that either of the two is used when expected? That would mean e.g. running a knn vectory query, with the executor set to the searcher, and checking that rewrite creates one task per segment, while collection does not.